File: project files/server/src/services/paymentService.js
const AppError = require('../utils/AppError');
const { ORDER_STATUS, PAYMENT_STATUS } = require('../utils/constants');
const orderRepository = require('../repositories/orderRepository');
const paymentRepository = require('../repositories/paymentRepository');
const cartRepository = require('../repositories/cartRepository');
const orderService = require('./orderService');
const paymentGatewayService = require('./paymentGatewayService');
const inventoryService = require('./inventoryService');
const processedWebhookEventRepository = require('../repositories/processedWebhookEventRepository');

const isPayableOnlineOrder = (order) =>
  order &&
  order.paymentMethod === 'ONLINE' &&
  order.orderStatus === ORDER_STATUS.PENDING_PAYMENT &&
  order.paymentStatus === PAYMENT_STATUS.PENDING;

const resolvePayableOrdersForRequest = async ({ order, userId }) => {
  if (String(order.userId) !== String(userId)) throw new AppError('Forbidden', 403, 'FORBIDDEN');
  if (order.paymentMethod !== 'ONLINE') throw new AppError('Order is not online payment type', 400, 'PAYMENT_MODE_INVALID');

  if (!order.orderGroupId) {
    if (!isPayableOnlineOrder(order)) {
      throw new AppError('Order not in payable state', 400, 'ORDER_STATE_INVALID');
    }
    return [order];
  }

  const children = await orderRepository.listByGroupId(order.orderGroupId);
  const sameUserChildren = children.filter((row) => String(row.userId) === String(userId));
  const payableChildren = sameUserChildren.filter((row) => isPayableOnlineOrder(row));
  if (!payableChildren.length) {
    throw new AppError('Order group has no payable child orders', 400, 'ORDER_STATE_INVALID');
  }
  return payableChildren;
};

const createGatewayOrder = async ({ orderId, userId, idempotencyKey }) => {
  const normalizedIdempotencyKey = String(idempotencyKey || '').trim();
  if (!normalizedIdempotencyKey) {
    throw new AppError('idempotencyKey is required', 400, 'IDEMPOTENCY_KEY_REQUIRED');
  }

  const existingTx = await paymentRepository.findByIdempotencyKey(normalizedIdempotencyKey);
  if (existingTx) {
    const withinWindow = new Date(existingTx.createdAt).getTime() >= Date.now() - 24 * 60 * 60 * 1000;
    if (withinWindow) {
      const relatedOrder = await orderRepository.findById(existingTx.orderId);
      return {
        paymentTransactionId: existingTx._id,
        orderId: relatedOrder?.orderId || orderId,
        externalOrderId: existingTx.externalOrderId,
        paymentUrl: existingTx.metadata?.paymentUrl || '',
        amount: existingTx.metadata?.payableAmount || existingTx.amount,
        currency: existingTx.currency
      };
    }
  }

  const requestedOrder = await orderRepository.findByOrderId(orderId);
  if (!requestedOrder) throw new AppError('Order not found', 404, 'ORDER_NOT_FOUND');

  const payableOrders = await resolvePayableOrdersForRequest({ order: requestedOrder, userId });
  const payableAmount = Number(payableOrders.reduce((sum, row) => sum + (row.totalAmount || 0), 0).toFixed(2));
  const gatewayReference = requestedOrder.orderGroupId || requestedOrder.orderId;

  const gatewayOrder = await paymentGatewayService.createPaymentOrder({
    orderId: gatewayReference,
    amount: payableAmount,
    currency: 'INR'
  });

  return inventoryService.withTransaction(async (session) => {
    const createdTransactions = [];
    for (const row of payableOrders) {
      const paymentDoc = await paymentRepository.createTransaction(
        {
          orderId: row._id,
          externalOrderId: gatewayOrder.externalOrderId,
          amount: row.totalAmount,
          currency: 'INR',
          gateway: gatewayOrder.provider,
          status: 'CREATED',
          metadata: {
            orderId: row.orderId,
            orderGroupId: row.orderGroupId || '',
            payableAmount,
            paymentUrl: gatewayOrder.paymentUrl
          }
        },
        session
      );
      if (createdTransactions.length === 0) {
        paymentDoc.idempotencyKey = normalizedIdempotencyKey;
        paymentDoc.idempotencyExpiresAt = new Date(Date.now() + 24 * 60 * 60 * 1000);
        await paymentRepository.save(paymentDoc, session);
      }
      createdTransactions.push(paymentDoc);

      const orderDoc = await orderRepository.findById(row._id, null, session);
      orderDoc.paymentGatewayOrderId = gatewayOrder.externalOrderId;
      await orderRepository.save(orderDoc, session);
    }

    return {
      paymentTransactionId: createdTransactions[0]?._id || null,
      orderId: requestedOrder.orderId,
      externalOrderId: gatewayOrder.externalOrderId,
      paymentUrl: gatewayOrder.paymentUrl,
      amount: gatewayOrder.amount,
      currency: gatewayOrder.currency
    };
  });
};

const handlePaymentWebhook = async ({ payload, signature, rawBody }) => {
  const verified = await paymentGatewayService.verifyWebhook(payload, signature, rawBody);
  if (!verified.valid) throw new AppError('Invalid webhook signature', 400, 'INVALID_WEBHOOK_SIGNATURE');

  if (verified.eventId) {
    const existingEvent = await processedWebhookEventRepository.findByEventId(verified.eventId);
    if (existingEvent) return { ok: true, duplicate: true };
    try {
      await processedWebhookEventRepository.createEvent({
        eventId: verified.eventId,
        provider: String(process.env.PAYMENT_PROVIDER || 'MOCK'),
        expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000)
      });
    } catch (error) {
      if (error?.code === 11000) return { ok: true, duplicate: true };
      throw error;
    }
  }

  const tx = await paymentRepository.findByExternalOrderId(verified.externalOrderId);
  if (!tx && !verified.refundReferenceId && !verified.externalPaymentId) {
    throw new AppError('Payment transaction not found', 404, 'PAYMENT_NOT_FOUND');
  }

  if (verified.status === 'CAPTURED' || verified.event === 'payment.captured') {
    return inventoryService.withTransaction(async (session) => {
      const txDocs = await paymentRepository.listByExternalOrderId(verified.externalOrderId, session);
      if (!txDocs.length) throw new AppError('Payment transaction not found', 404, 'PAYMENT_NOT_FOUND');

      let hasPendingCapture = false;
      for (const txDoc of txDocs) {
        if (txDoc.status !== 'CAPTURED') {
          hasPendingCapture = true;
          txDoc.status = 'CAPTURED';
          txDoc.externalPaymentId = verified.externalPaymentId || txDoc.externalPaymentId;
          await paymentRepository.save(txDoc, session);
        }
      }
      if (!hasPendingCapture) return { ok: true, duplicate: true };

      let confirmedCount = 0;
      for (const txDoc of txDocs) {
        const orderDoc = await orderRepository.findById(txDoc.orderId, null, session);
        if (!orderDoc || orderDoc.orderStatus !== ORDER_STATUS.PENDING_PAYMENT) {
          continue;
        }
        await orderService.confirmPayment({
          orderId: orderDoc.orderId,
          paymentGatewayOrderId: verified.externalOrderId,
          paymentGatewayPaymentId: verified.externalPaymentId,
          updatedBy: orderDoc.userId,
          updatedByRole: 'ADMIN',
          session
        });
        confirmedCount += 1;
      }
      return { ok: true, confirmedCount };
    });
  }

  if (
    verified.status === 'REFUNDED' ||
    verified.event === 'refund.processed' ||
    verified.event === 'payment.refund.processed'
  ) {
    return inventoryService.withTransaction(async (session) => {
      let refundTx = null;
      if (verified.refundReferenceId) {
        refundTx = await paymentRepository.findByRefundReferenceId(verified.refundReferenceId, session);
      }
      if (!refundTx && verified.externalPaymentId) {
        refundTx = await paymentRepository.findByExternalPaymentId(verified.externalPaymentId, session);
      }
      if (!refundTx && tx) {
        refundTx = await paymentRepository.findById(tx._id, session);
      }
      if (!refundTx) throw new AppError('Payment transaction not found', 404, 'PAYMENT_NOT_FOUND');
      if (refundTx.status === 'REFUNDED') return { ok: true, duplicate: true };

      refundTx.status = 'REFUNDED';
      if (verified.refundReferenceId) {
        refundTx.refundReferenceId = verified.refundReferenceId;
      }
      await paymentRepository.save(refundTx, session);

      const orderDoc = await orderRepository.findById(refundTx.orderId, null, session);
      if (!orderDoc) throw new AppError('Order not found', 404, 'ORDER_NOT_FOUND');
      if (orderDoc.orderStatus === ORDER_STATUS.REFUNDED) return { ok: true, duplicate: true };
      if (orderDoc.orderStatus !== ORDER_STATUS.REFUND_INITIATED) {
        return { ok: true };
      }

      await orderService.transitionOrderStatus({
        orderId: orderDoc.orderId,
        nextStatus: ORDER_STATUS.REFUNDED,
        updatedBy: orderDoc.userId,
        updatedByRole: 'ADMIN',
        session
      });
      return { ok: true };
    });
  }

  if (verified.status === 'FAILED') {
    return inventoryService.withTransaction(async (session) => {
      const txDocs = await paymentRepository.listByExternalOrderId(verified.externalOrderId, session);
      if (!txDocs.length) throw new AppError('Payment transaction not found', 404, 'PAYMENT_NOT_FOUND');
      const pending = txDocs.filter((row) => row.status !== 'FAILED');
      if (!pending.length) return { ok: true, duplicate: true };
      const restoredUsers = new Set();
      for (const txDoc of pending) {
        txDoc.status = 'FAILED';
        await paymentRepository.save(txDoc, session);
        const orderDoc = await orderRepository.findById(txDoc.orderId, null, session);
        if (!orderDoc) continue;
        if (orderDoc.orderStatus !== ORDER_STATUS.PENDING_PAYMENT) continue;
        const userKey = String(orderDoc.userId);
        if (restoredUsers.has(userKey)) continue;
        restoredUsers.add(userKey);
        await cartRepository.mergeItems({
          userId: orderDoc.userId,
          items: (orderDoc.orderItems || []).map((item) => ({
            productId: item.productId,
            variantId: item.variantId,
            quantity: item.quantity
          })),
          couponCode: orderDoc.couponCode || '',
          session
        });
      }
      return { ok: true };
    });
  }

  return { ok: true };
};

const initiateRefund = async ({ orderId, requestedBy }) => {
  const order = await orderRepository.findByOrderId(orderId);
  if (!order) throw new AppError('Order not found', 404, 'ORDER_NOT_FOUND');

  const tx = await paymentRepository.findByExternalOrderId(order.paymentGatewayOrderId);
  if (!tx) throw new AppError('Payment transaction not found', 404, 'PAYMENT_NOT_FOUND');
  const refund = await paymentGatewayService.initiateRefund({
    externalPaymentId: tx.externalPaymentId,
    amount: order.totalAmount
  });

  return inventoryService.withTransaction(async (session) => {
    const txDoc = await paymentRepository.findById(tx._id, session);
    txDoc.status = 'REFUND_INITIATED';
    txDoc.refundReferenceId = refund.refundReferenceId;
    await paymentRepository.save(txDoc, session);

    const orderDoc = await orderRepository.findById(order._id, null, session);
    orderDoc.refundReferenceId = refund.refundReferenceId;
    await orderRepository.save(orderDoc, session);

    await orderService.transitionOrderStatus({
      orderId: order.orderId,
      nextStatus: 'REFUND_INITIATED',
      updatedBy: requestedBy,
      updatedByRole: 'ADMIN',
      session
    });

    return {
      orderId: order.orderId,
      refundReferenceId: refund.refundReferenceId
    };
  });
};

module.exports = {
  createGatewayOrder,
  handlePaymentWebhook,
  initiateRefund
};

File: spring_back/src/main/java/com/groceryapp/service/PaymentService.java
package com.groceryapp.service;

import com.groceryapp.config.AppProperties;
import com.groceryapp.exception.AppException;
import com.groceryapp.model.*;
import com.groceryapp.repository.*;
import com.groceryapp.util.OrderStatus;
import com.groceryapp.util.PaymentStatus;
import com.groceryapp.util.Roles;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.*;

@Service
public class PaymentService {
    private final OrderRepository orderRepository;
    private final PaymentTransactionRepository paymentTransactionRepository;
    private final CartRepository cartRepository;
    private final OrderService orderService;
    private final PaymentGatewayService paymentGatewayService;
    private final ProcessedWebhookEventRepository processedWebhookEventRepository;
    private final AppProperties appProperties;

    public PaymentService(
            OrderRepository orderRepository,
            PaymentTransactionRepository paymentTransactionRepository,
            CartRepository cartRepository,
            OrderService orderService,
            PaymentGatewayService paymentGatewayService,
            ProcessedWebhookEventRepository processedWebhookEventRepository,
            AppProperties appProperties
    ) {
        this.orderRepository = orderRepository;
        this.paymentTransactionRepository = paymentTransactionRepository;
        this.cartRepository = cartRepository;
        this.orderService = orderService;
        this.paymentGatewayService = paymentGatewayService;
        this.processedWebhookEventRepository = processedWebhookEventRepository;
        this.appProperties = appProperties;
    }

    public Map<String, Object> createGatewayOrder(String orderId, String userId, String idempotencyKey) {
        String normalizedIdempotencyKey = str(idempotencyKey);
        if (normalizedIdempotencyKey.isBlank()) {
            throw new AppException("idempotencyKey is required", 400, "IDEMPOTENCY_KEY_REQUIRED");
        }

        PaymentTransaction existingTx = paymentTransactionRepository.findByIdempotencyKey(normalizedIdempotencyKey).orElse(null);
        if (existingTx != null && existingTx.getCreatedAt() != null && existingTx.getCreatedAt().isAfter(Instant.now().minusSeconds(24 * 60 * 60))) {
            Order relatedOrder = orderRepository.findById(existingTx.getOrderId()).orElse(null);
            Map<String, Object> meta = existingTx.getMetadata() == null ? Map.of() : existingTx.getMetadata();
            return Map.of(
                    "paymentTransactionId", existingTx.getId(),
                    "orderId", relatedOrder == null ? orderId : relatedOrder.getOrderId(),
                    "externalOrderId", str(existingTx.getExternalOrderId()),
                    "paymentUrl", str(meta.get("paymentUrl")),
                    "amount", num(meta.get("payableAmount"), num(existingTx.getAmount())),
                    "currency", str(existingTx.getCurrency())
            );
        }

        Order requestedOrder = orderRepository.findByOrderId(orderId)
                .orElseThrow(() -> new AppException("Order not found", 404, "ORDER_NOT_FOUND"));
        List<Order> payableOrders = resolvePayableOrdersForRequest(requestedOrder, userId);
        double payableAmount = round2(payableOrders.stream().mapToDouble(o -> num(o.getTotalAmount())).sum());
        String gatewayReference = blank(requestedOrder.getOrderGroupId()) ? requestedOrder.getOrderId() : requestedOrder.getOrderGroupId();
        Map<String, Object> gatewayOrder = paymentGatewayService.createPaymentOrder(gatewayReference, payableAmount, "INR");
        String externalOrderId = str(gatewayOrder.get("externalOrderId"));

        List<PaymentTransaction> created = new ArrayList<>();
        for (Order row : payableOrders) {
            PaymentTransaction tx = PaymentTransaction.builder()
                    .orderId(row.getId())
                    .externalOrderId(externalOrderId)
                    .amount(num(row.getTotalAmount()))
                    .currency("INR")
                    .gateway(str(gatewayOrder.get("provider")))
                    .status("CREATED")
                    .metadata(Map.of(
                            "orderId", row.getOrderId(),
                            "orderGroupId", blank(row.getOrderGroupId()) ? "" : row.getOrderGroupId(),
                            "payableAmount", payableAmount,
                            "paymentUrl", str(gatewayOrder.get("paymentUrl"))
                    ))
                    .createdAt(Instant.now())
                    .updatedAt(Instant.now())
                    .build();
            if (created.isEmpty()) {
                tx.setIdempotencyKey(normalizedIdempotencyKey);
                tx.setIdempotencyExpiresAt(Instant.now().plusSeconds(24 * 60 * 60));
            }
            tx = paymentTransactionRepository.save(tx);
            created.add(tx);

            Order doc = orderRepository.findById(row.getId()).orElse(null);
            if (doc != null) {
                doc.setPaymentGatewayOrderId(externalOrderId);
                doc.setUpdatedAt(Instant.now());
                orderRepository.save(doc);
            }
        }

        return Map.of(
                "paymentTransactionId", created.isEmpty() ? null : created.get(0).getId(),
                "orderId", requestedOrder.getOrderId(),
                "externalOrderId", externalOrderId,
                "paymentUrl", str(gatewayOrder.get("paymentUrl")),
                "amount", num(gatewayOrder.get("amount"), payableAmount),
                "currency", str(gatewayOrder.get("currency"))
        );
    }

    public Map<String, Object> handlePaymentWebhook(Map<String, Object> payload, String signature, byte[] rawBody) {
        PaymentGatewayService.VerifiedWebhook verified = paymentGatewayService.verifyWebhook(payload, signature, rawBody);
        if (!verified.valid()) {
            throw new AppException("Invalid webhook signature", 400, "INVALID_WEBHOOK_SIGNATURE");
        }

        if (!blank(verified.eventId())) {
            ProcessedWebhookEvent existingEvent = processedWebhookEventRepository.findByEventId(verified.eventId()).orElse(null);
            if (existingEvent != null) {
                return Map.of("ok", true, "duplicate", true);
            }
            try {
                processedWebhookEventRepository.save(ProcessedWebhookEvent.builder()
                        .eventId(verified.eventId())
                        .provider(appProperties.getPaymentProvider())
                        .expiresAt(Instant.now().plusSeconds(7L * 24 * 60 * 60))
                        .createdAt(Instant.now())
                        .updatedAt(Instant.now())
                        .build());
            } catch (Exception ignored) {
                return Map.of("ok", true, "duplicate", true);
            }
        }

        PaymentTransaction tx = paymentTransactionRepository.findFirstByExternalOrderId(verified.externalOrderId()).orElse(null);
        if (tx == null && blank(verified.refundReferenceId()) && blank(verified.externalPaymentId())) {
            throw new AppException("Payment transaction not found", 404, "PAYMENT_NOT_FOUND");
        }

        if ("CAPTURED".equalsIgnoreCase(verified.status()) || "payment.captured".equalsIgnoreCase(verified.event())) {
            List<PaymentTransaction> txDocs = paymentTransactionRepository.findByExternalOrderId(verified.externalOrderId());
            if (txDocs.isEmpty()) {
                throw new AppException("Payment transaction not found", 404, "PAYMENT_NOT_FOUND");
            }
            boolean hasPendingCapture = false;
            for (PaymentTransaction txDoc : txDocs) {
                if (!"CAPTURED".equalsIgnoreCase(txDoc.getStatus())) {
                    hasPendingCapture = true;
                    txDoc.setStatus("CAPTURED");
                    if (!blank(verified.externalPaymentId())) {
                        txDoc.setExternalPaymentId(verified.externalPaymentId());
                    }
                    txDoc.setUpdatedAt(Instant.now());
                    paymentTransactionRepository.save(txDoc);
                }
            }
            if (!hasPendingCapture) {
                return Map.of("ok", true, "duplicate", true);
            }

            int confirmedCount = 0;
            for (PaymentTransaction txDoc : txDocs) {
                Order orderDoc = orderRepository.findById(txDoc.getOrderId()).orElse(null);
                if (orderDoc == null || !OrderStatus.PENDING_PAYMENT.equals(orderDoc.getOrderStatus())) {
                    continue;
                }
                orderService.confirmPayment(
                        orderDoc.getOrderId(),
                        verified.externalOrderId(),
                        verified.externalPaymentId(),
                        orderDoc.getUserId(),
                        Roles.ADMIN
                );
                confirmedCount++;
            }
            return Map.of("ok", true, "confirmedCount", confirmedCount);
        }

        if ("REFUNDED".equalsIgnoreCase(verified.status())
                || "refund.processed".equalsIgnoreCase(verified.event())
                || "payment.refund.processed".equalsIgnoreCase(verified.event())) {
            PaymentTransaction refundTx = null;
            if (!blank(verified.refundReferenceId())) {
                refundTx = paymentTransactionRepository.findByRefundReferenceId(verified.refundReferenceId()).orElse(null);
            }
            if (refundTx == null && !blank(verified.externalPaymentId())) {
                refundTx = paymentTransactionRepository.findByExternalPaymentId(verified.externalPaymentId()).orElse(null);
            }
            if (refundTx == null && tx != null) {
                refundTx = paymentTransactionRepository.findById(tx.getId()).orElse(null);
            }
            if (refundTx == null) {
                throw new AppException("Payment transaction not found", 404, "PAYMENT_NOT_FOUND");
            }
            if ("REFUNDED".equalsIgnoreCase(refundTx.getStatus())) {
                return Map.of("ok", true, "duplicate", true);
            }

            refundTx.setStatus("REFUNDED");
            if (!blank(verified.refundReferenceId())) {
                refundTx.setRefundReferenceId(verified.refundReferenceId());
            }
            refundTx.setUpdatedAt(Instant.now());
            paymentTransactionRepository.save(refundTx);

            Order orderDoc = orderRepository.findById(refundTx.getOrderId())
                    .orElseThrow(() -> new AppException("Order not found", 404, "ORDER_NOT_FOUND"));
            if (OrderStatus.REFUNDED.equals(orderDoc.getOrderStatus())) {
                return Map.of("ok", true, "duplicate", true);
            }
            if (!OrderStatus.REFUND_INITIATED.equals(orderDoc.getOrderStatus())) {
                return Map.of("ok", true);
            }
            orderService.transitionOrderStatus(orderDoc.getOrderId(), OrderStatus.REFUNDED, orderDoc.getUserId(), Roles.ADMIN, null, null);
            return Map.of("ok", true);
        }

        if ("FAILED".equalsIgnoreCase(verified.status())) {
            List<PaymentTransaction> txDocs = paymentTransactionRepository.findByExternalOrderId(verified.externalOrderId());
            if (txDocs.isEmpty()) {
                throw new AppException("Payment transaction not found", 404, "PAYMENT_NOT_FOUND");
            }
            List<PaymentTransaction> pending = txDocs.stream().filter(row -> !"FAILED".equalsIgnoreCase(row.getStatus())).toList();
            if (pending.isEmpty()) {
                return Map.of("ok", true, "duplicate", true);
            }
            Set<String> restoredUsers = new HashSet<>();
            for (PaymentTransaction txDoc : pending) {
                txDoc.setStatus("FAILED");
                txDoc.setUpdatedAt(Instant.now());
                paymentTransactionRepository.save(txDoc);

                Order orderDoc = orderRepository.findById(txDoc.getOrderId()).orElse(null);
                if (orderDoc == null || !OrderStatus.PENDING_PAYMENT.equals(orderDoc.getOrderStatus())) continue;

                String userKey = str(orderDoc.getUserId());
                if (restoredUsers.contains(userKey)) continue;
                restoredUsers.add(userKey);
                mergeOrderItemsToCart(orderDoc);
            }
            return Map.of("ok", true);
        }

        return Map.of("ok", true);
    }

    public Map<String, Object> initiateRefund(String orderId, String requestedBy) {
        Order order = orderRepository.findByOrderId(orderId)
                .orElseThrow(() -> new AppException("Order not found", 404, "ORDER_NOT_FOUND"));
        PaymentTransaction tx = paymentTransactionRepository.findFirstByExternalOrderId(order.getPaymentGatewayOrderId())
                .orElseThrow(() -> new AppException("Payment transaction not found", 404, "PAYMENT_NOT_FOUND"));
        Map<String, Object> refund = paymentGatewayService.initiateRefund(tx.getExternalPaymentId(), num(order.getTotalAmount()));

        tx.setStatus("REFUND_INITIATED");
        tx.setRefundReferenceId(str(refund.get("refundReferenceId")));
        tx.setUpdatedAt(Instant.now());
        paymentTransactionRepository.save(tx);

        Order orderDoc = orderRepository.findById(order.getId()).orElseThrow(() -> new AppException("Order not found", 404, "ORDER_NOT_FOUND"));
        orderDoc.setRefundReferenceId(str(refund.get("refundReferenceId")));
        orderDoc.setUpdatedAt(Instant.now());
        orderRepository.save(orderDoc);

        orderService.transitionOrderStatus(order.getOrderId(), OrderStatus.REFUND_INITIATED, requestedBy, Roles.ADMIN, null, null);
        return Map.of("orderId", order.getOrderId(), "refundReferenceId", str(refund.get("refundReferenceId")));
    }

    private List<Order> resolvePayableOrdersForRequest(Order order, String userId) {
        if (!Objects.equals(str(order.getUserId()), str(userId))) {
            throw new AppException("Forbidden", 403, "FORBIDDEN");
        }
        if (!"ONLINE".equalsIgnoreCase(order.getPaymentMethod())) {
            throw new AppException("Order is not online payment type", 400, "PAYMENT_MODE_INVALID");
        }
        if (blank(order.getOrderGroupId())) {
            if (!isPayableOnlineOrder(order)) {
                throw new AppException("Order not in payable state", 400, "ORDER_STATE_INVALID");
            }
            return List.of(order);
        }
        List<Order> children = orderRepository.findByOrderGroupIdOrderByCreatedAtAsc(order.getOrderGroupId());
        List<Order> sameUserChildren = children.stream().filter(row -> Objects.equals(str(row.getUserId()), str(userId))).toList();
        List<Order> payableChildren = sameUserChildren.stream().filter(this::isPayableOnlineOrder).toList();
        if (payableChildren.isEmpty()) {
            throw new AppException("Order group has no payable child orders", 400, "ORDER_STATE_INVALID");
        }
        return payableChildren;
    }

    private boolean isPayableOnlineOrder(Order order) {
        return order != null
                && "ONLINE".equalsIgnoreCase(order.getPaymentMethod())
                && OrderStatus.PENDING_PAYMENT.equals(order.getOrderStatus())
                && PaymentStatus.PENDING.equals(order.getPaymentStatus());
    }

    private void mergeOrderItemsToCart(Order orderDoc) {
        Cart cart = cartRepository.findByUserId(orderDoc.getUserId()).orElseGet(() -> Cart.builder()
                .userId(orderDoc.getUserId())
                .items(new ArrayList<>())
                .couponCode("")
                .createdAt(Instant.now())
                .updatedAt(Instant.now())
                .build());
        List<Cart.CartItem> items = cart.getItems() == null ? new ArrayList<>() : new ArrayList<>(cart.getItems());
        for (Order.OrderItem item : (orderDoc.getOrderItems() == null ? List.<Order.OrderItem>of() : orderDoc.getOrderItems())) {
            int quantity = item.getQuantity() == null ? 0 : item.getQuantity();
            if (quantity <= 0) continue;
            Cart.CartItem existing = items.stream()
                    .filter(row -> Objects.equals(row.getProductId(), item.getProductId()) && Objects.equals(row.getVariantId(), item.getVariantId()))
                    .findFirst().orElse(null);
            if (existing != null) {
                existing.setQuantity((existing.getQuantity() == null ? 0 : existing.getQuantity()) + quantity);
            } else {
                items.add(Cart.CartItem.builder()
                        .productId(item.getProductId())
                        .variantId(item.getVariantId())
                        .quantity(quantity)
                        .build());
            }
        }
        cart.setItems(items);
        if (!blank(orderDoc.getCouponCode()) && blank(cart.getCouponCode())) {
            cart.setCouponCode(orderDoc.getCouponCode().toUpperCase(Locale.ROOT));
        }
        cart.setUpdatedAt(Instant.now());
        cartRepository.save(cart);
    }

    private String str(Object value) {
        return value == null ? "" : String.valueOf(value).trim();
    }

    private boolean blank(String value) {
        return value == null || value.trim().isBlank();
    }

    private double num(Double value) {
        return value == null ? 0d : value;
    }

    private double num(Object value, double def) {
        try {
            return value == null ? def : Double.parseDouble(String.valueOf(value));
        } catch (Exception ex) {
            return def;
        }
    }

    private double round2(double val) {
        return Math.round(val * 100.0) / 100.0;
    }
}
